### ISSUE TITLE
Resource Server related FAILURE : received access token verification

### ISSUE THREAD

**Issue by tnorimat** [jsoss-sig/keycloak-fapi#27]
Some test module like `fapi-rw-id2-ensure-request-object-with-multiple-aud-succeeds-with-mtls` and `fapi-rw-id2-ensure-request-object-with-multiple-aud-succeeds-with-private-key-and-mtls-holder-of-key` checks whether a resource server returns error when it receives invalid access token.
 To pass these tests, the resource server verify whether received access token is valid or not by token introspection and returns appropriate response.

**Comment by wadahiro** [comment]
On [fix-resource-server](https://github.com/jsoss-sig/keycloak-fapi/commits/fix-resource-server) branch, I implemented the introspection and MTLS verification on the resource server temporarily by keycloak-gatekeeper customization in the following commit:

- Add token introspection: https://github.com/openstandia/keycloak-gatekeeper/commit/7f65d355f709b421f3cb2596c310aa3fcd359da5
- Add MTLS verification with token introspection: https://github.com/openstandia/keycloak-gatekeeper/commit/878b020f41057cf327fc7accd10950e66f018dd8

Then I tested `fapi-rw-id2-ensure-request-object-with-multiple-aud-succeeds-with-private-key-and-mtls-holder-of-key` again.

But `CallAccountsEndpointWithBearerTokenExpectingError` (RFC6749-4.1.2) wasn't passed correctly due to WARN. The test-case attempts reuse of client2's authorization code & test if access token is revoked:

![image](https://user-images.githubusercontent.com/28739/82964776-47e7b380-a001-11ea-9096-a1d4dbd5e1e6.png)

I checked RFC6749-4.1.2. It says:

```
         if an authorization code is used more than
         once, the authorization server MUST deny the request and SHOULD
         revoke (when possible) all tokens previously issued based on
         that authorization code.
```

Should we support the revocation in keycloak? Or it's implemented on the latest keycloak?

**Comment by wadahiro** [comment]
I looked keycloak code: https://github.com/keycloak/keycloak/blob/6.0.1/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L286-L289

It seems that keycloak invalidates existing clientSession. It is same code on the latest keycloak.
There may be a problem on the token introspection side...?

**Comment by wadahiro** [comment]
After upgrading to the latest conformance suite ([12b18ce5](https://gitlab.com/openid/conformance-suite/-/tree/12b18ce53b3a1ea7a632854143fc385488a1c2ef)), `CallProtectedResourceWithBearerTokenExpectingError` which was renamed from  
 `CallAccountsEndpointWithBearerTokenExpectingError ` in [this commit](https://gitlab.com/openid/conformance-suite/-/commit/887b1c6b63b06bbdc60c3c0c5456cd7fd75bf069) passed.

In the previous version, the test reused the same JWT for client authentication which is mentioned in https://github.com/jsoss-sig/keycloak-fapi/issues/10. It returns ["Token reuse detected" error](https://github.com/keycloak/keycloak/blob/0d0617d44af197e627af378438357e2d43242a69/services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientAuthenticator.java#L182) when the client called token endpoint. That's why the access token wasn't revoked.

In the latest version,  the test doesn't reuse it. So the access token is revoked expectedly.

 In conclusion, we can close this issue now.

**Comment by wadahiro** [comment]
Also, @tnorimat confirmed the introspection endpoint returns `active=false` expectedly in the keycloak Arquillian Integration Test. There is no problem with keycloak side.

**Comment by tnorimat** [comment]
@wadahiro Thank you. I'll close this issue.
